kv/bulk: compute accurate stats in StreamSSTBatcher

msbutler opened the pull request:
This patch teaches PCR to compute an accurate stats diff on its ingestion
path.

Fixes https://github.com/cockroachdb/cockroach/issues/145548

Release note: none

msbutler committed:
roachtest: run c2c/kv0 with zipfian distribution

The zipfian distribution tests PCR ingestion of an update heavy workload.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149923)
<!-- Reviewable:end -->

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 8.422m ±0% | 8.443m ±0% | +0.25% | p=0.010 n=15    |
| ⚪ **allocs/op** | 8.956k ±0% | 8.972k ±0% | ~ | p=0.076 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/c952e8a/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/c952e8a202cafae8c8c66fbd5dde8fcbe399e81f/bin/pkg_sql_tests benchdiff/c952e8a/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/c952e8a/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/8fefbf9/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8fefbf9368cdac249e02e9ffd6081cf026e9b788/bin/pkg_sql_tests benchdiff/8fefbf9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8fefbf9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=8fefbf9 --new=c952e8a ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>🔴 Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| 🔴 **sec/op** | 1.072m ±1% | 1.082m ±1% | +0.98% | p=0.001 n=15    |
| ⚪ **allocs/op** | 1.859k ±0% | 1.860k ±0% | +0.05% | p=0.020 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/c952e8a/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/c952e8a202cafae8c8c66fbd5dde8fcbe399e81f/bin/pkg_sql_tests benchdiff/c952e8a/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/c952e8a/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/8fefbf9/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8fefbf9368cdac249e02e9ffd6081cf026e9b788/bin/pkg_sql_tests benchdiff/8fefbf9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8fefbf9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=8fefbf9 --new=c952e8a ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 1.941m ±0% | 1.945m ±0% | ~ | p=0.067 n=15    |
| ⚪ **allocs/op** | 3.805k ±0% | 3.808k ±0% | ~ | p=0.169 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/c952e8a/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/c952e8a202cafae8c8c66fbd5dde8fcbe399e81f/bin/pkg_sql_tests benchdiff/c952e8a/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/c952e8a/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/8fefbf9/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8fefbf9368cdac249e02e9ffd6081cf026e9b788/bin/pkg_sql_tests benchdiff/8fefbf9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8fefbf9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=8fefbf9 --new=c952e8a ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/c952e8a202cafae8c8c66fbd5dde8fcbe399e81f/16220058590-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/8fefbf9368cdac249e02e9ffd6081cf026e9b788/16220058590-1/\* old/
```

</details>

_built with commit: [c952e8a202cafae8c8c66fbd5dde8fcbe399e81f](https://github.com/cockroachdb/cockroach/commit/c952e8a202cafae8c8c66fbd5dde8fcbe399e81f)_

<!-- microbench-ci -->

msbutler committed:
kv/bulk: compute accurate stats in StreamSSTBatcher

This patch teaches PCR to compute an accurate stats diff on its ingestion
path.

Fixes #145548

Release note: none

jeffswenson reviewed (approved):
LGTM

Is there an e2e test for this functionality? E.g. maybe a test that sets up PCR, writes some random traffic, then asserts the stats are correct.

msbutler commented:
> Is there an e2e test for this functionality? 

Done. That took longer than i'd like to admin :D

jeffswenson reviewed (approved):
LGTM

Test looks great 💯

msbutler commented:
TFTR!

bors r=jeffswenson
